classifier/run_classifier_utils.py

```diff
@@ -85,68 +61,74 @@ def __init__(self, config, num_tags,
       use_one_hot_embeddings: Whether to use one-hot embeddings for word
         embeddings.
       max_seq_length: Maximum sequence length.
+      classifier_type: Either Grammar or Meaning.
     """
     self._config = config
-    self._num_tags = num_tags
+    self._num_categories = num_categories
     self._init_checkpoint = init_checkpoint
     self._learning_rate = learning_rate
     self._num_train_steps = num_train_steps
     self._num_warmup_steps = num_warmup_steps
     self._use_tpu = use_tpu
     self._use_one_hot_embeddings = use_one_hot_embeddings
     self._max_seq_length = max_seq_length
+    self._classifier_type = classifier_type
 
-  def _create_model(self, mode, input_ids, input_mask, segment_ids, labels,
-                    labels_mask):
+  def _create_model(self, mode, input_ids_source, input_mask_source, 
+                    segment_ids_source, input_ids_summary,
+                    input_mask_summary, segment_ids_summary, labels):
     """Creates a LaserTagger model."""
     is_training = (mode == tf.estimator.ModeKeys.TRAIN)
-    model = modeling.BertModel(
+    model_source = modeling.BertModel(
         config=self._config,
         is_training=is_training,
-        input_ids=input_ids,
-        input_mask=input_mask,
-        token_type_ids=segment_ids,
+        input_ids=input_ids_source,
+        input_mask=input_mask_source,
+        token_type_ids=segment_ids_source,
         use_one_hot_embeddings=self._use_one_hot_embeddings)
-
-    final_hidden = model.get_sequence_output()
-
-    if self._config.use_t2t_decoder:
-      # Size of the output vocabulary which contains the tags + begin and end
-      # tokens used by the Transformer decoder.
-      output_vocab_size = self._num_tags + 2
-      params = _get_decoder_params(self._config, self._use_tpu,
-                                   self._max_seq_length, output_vocab_size)
-      decoder = transformer_decoder.TransformerDecoder(params, is_training)
-      logits = decoder(input_mask, final_hidden, labels)
+    final_hidden_source = model_source.get_sequence_output()
+    
+    if self._classifier_type == "Meaning":
```

Comment:
QQ: I'm wondering why you choose to train two BERT models and concatenate them here?

Reply:
Bert learns the context of the input text, and I do not want the context of either the source or the summary to affect the embedding of the other. This is my reasoning for this decision. What would you suggest doing here? Thanks!

Reply:
Are input_ids_source and input_ids_summary aligned? (I think so since you can concatenate them in the output)
If so usually people concatenate their embedding first then feed into the BERT model.

Reply:
I'm not sure what you mean by aligned? They are one pair of source and summary, but they are of different length usually. They are padded to the same length and then fed into BERT separately. If I directly concatenate them, I was worried that they would be treated like one continuous piece of text. Particularly, from what I read online, the first token in the output of BERT kind of summarizes the meaning of the input text. If I feed source and summary separately, they would each get such a token in the embedding, which would be kind of useful in this case because we hope that the classifier can learn how well the meaning of the summary matches that of the source. But if I concatenate them, then there would only be one such token representing the meaning of both source and summary, and I am not sure whether that would impact the performance.

Reply:
I see, you get the first output of the BERT for both. Then it makes sense to me. Thanks for the explanation!